rust-mode.el

```diff
@@ -267,18 +268,40 @@ to the function arguments.  When nil, `->' will be indented one level."
         ;; Rewind until the point no longer moves
         (setq continue (/= starting (point)))))))
 
-(defun rust-in-macro ()
+(defvar-local rust-macro-scopes nil)
+
+(defun rust-macro-scope (&optional start end)
   (save-excursion
-    (when (> (rust-paren-level) 0)
-      (backward-up-list)
-      (rust-rewind-irrelevant)
-      (or (rust-looking-back-macro)
-          (and (rust-looking-back-ident)
-               (save-excursion
-                 (backward-sexp)
-                 (rust-rewind-irrelevant)
-                 (rust-looking-back-str "macro_rules!")))
-          (rust-in-macro)))))
+    ;; need to special case macro_rules which has unique syntax
+    (let ((scope nil))
+      (goto-char
+       (or start (point-min)))
+      ;; if there is a start move back to the first top level
+      (when start
+        (let ((top (syntax-ppss-toplevel-pos (syntax-ppss))))
+          (when top
+            (goto-char top))))
+      (while (search-forward "!" end t)
+        (when (and (not (rust-in-str-or-cmnt))
+                   (skip-chars-forward " \t\n\r")
+                   (or
+                    (eq ?\[ (char-after))
+                    (eq ?\( (char-after))
+                    (eq ?\{ (char-after))))
+          (let ((start (point)))
+            (forward-list)
+            (setq scope (cons (list start (point)) scope)))))
+      scope)))
+
+(defun rust-in-macro (&optional start end)
+  (when (> (rust-paren-level) 0)
+    (let ((scopes
+           (or rust-macro-scopes (rust-macro-scope start end))))
+      (seq-some
+       (lambda (sc)
+         (and (>= (point) (car sc))
+              (<= (point) (cadr sc))))
+       scopes))))
```

Comment:
I don't think `seq-some` is supported in emacs 24. And anyway I think using a `dolist` with `catch`+`throw` would allow us to exit early assuming `scopes` are ordered.
```elisp
(catch 'break
  (dolist (sc the-scopes)
    (if (< (point) (car sc)) (throw 'break nil))
    (if (<= (point) (cadr sc)) (throw 'break t))))
```

Reply:
Or the other way round, I might have assumed the wrong `scopes` order

Reply:
`scopes` is backward. The original plan I had it wasn't going to be ordered either.

If `seq-some` is supported in Emacs-24, it will be via ELPA. But, Emacs-27 is now in release-candidate. Do we really need to support Emacs-24?

Reply:
I don't have an opinion on supporting 24, but the github actions tests were breaking for me, so I guessed unless the maintainers decide to bump the requirement and remove that test, it needs to support 24.

https://github.com/phillord/rust-mode/actions/runs/92635709